src/mlpack/core/cv/cv_base.hpp

```diff
@@ -0,0 +1,329 @@
+ * numClasses) and to assert that the machine learning algorithm and data
+ * satisfy certain conditions.
+ *
+ * @tparam MLAlgorithm A machine learning algorithm.
+ * @tparam MatType The type of data.
+ * @tparam PredictionsType The type of predictions (labels/responses).
+ * @tparam WeightsType The type of weights. It supposed to be void* when weights
+ *     are not supported.
+ */
+template<typename MLAlgorithm,
+         typename MatType,
+         typename PredictionsType,
+         typename WeightsType>
+class CVBase
+{
+ public:
+  /*
+   * This constructor can be used for regression algorithms and for binary
+   * classification algorithms.
+   *
+   * @param xs Dataset to cross-validate on.
+   * @param ys Predictions (labels for classification algorithms and responses
+   *     for regression algorithms) for each point from the dataset.
+   *
+   * @tparam MT A type that can be converted to MatType.
+   * @tparam PT A type that can be converted to PredictionsType.
+   */
+  template<typename MT, typename PT>
+  CVBase(const MT& xs,
+         const PT& ys);
+
+  /*
+   * This constructor can be used for multiclass classification algorithms.
+   *
+   * @param xs Dataset to cross-validate on.
+   * @param ys Labels for each point from the dataset.
+   * @param numClasses Number of classes in the dataset.
+   *
+   * @tparam MT A type that can be converted to MatType.
+   * @tparam PT A type that can be converted to PredictionsType.
+   */
+  template<typename MT, typename PT>
+  CVBase(const MT& xs,
+         const PT& ys,
+         const size_t numClasses);
+
+  /*
+   * This constructor can be used for multiclass classification algorithms that
+   * can take a data::DatasetInfo parameter.
+   *
+   * @param xs Dataset to cross-validate on.
+   * @param datasetInfo Type information for each dimension of the dataset.
+   * @param ys Labels for each point from the dataset.
+   * @param numClasses Number of classes in the dataset.
+   *
+   * @tparam MT A type that can be converted to MatType.
+   * @tparam PT A type that can be converted to PredictionsType.
+   */
+  template<typename MT, typename PT>
+  CVBase(const MT& xs,
+         const data::DatasetInfo& datasetInfo,
+         const PT& ys,
+         const size_t numClasses);
+
+  /*
+   * This constructor can be used for regression and binary classification
+   * algorithms that support weighted learning.
+   *
+   * @param xs Dataset to cross-validate on.
+   * @param ys Predictions (labels for classification algorithms and responses
+   *     for regression algorithms) for each point from the dataset.
+   * @param weights Observation weights (for boosting).
+   *
+   * @tparam MT A type that can be converted to MatType.
+   * @tparam PT A type that can be converted to PredictionsType.
+   * @tparam WT A type that can be converted to WeightsType.
+   */
+  template<typename MT, typename PT, typename WT>
+  CVBase(const MT& xs,
+         const PT& ys,
+         const WT& weights);
```

Comment:
For these, does it make sense to use a `std::enable_if<>` here?  I guess that if the wrong overload is used a compiler error will still happen anyway, but maybe it will be trickier to diagnose.